core/controllers/async_download.py

```diff
@@ -0,0 +1,112 @@
+import io
+import uuid
+from typing import IO
+
+from flask import current_app
+from notifications_python_client.notifications import NotificationsAPIClient
+from werkzeug.datastructures import FileStorage
+
+from config import Config
+from core.aws import _S3_CLIENT
+from core.controllers.download import download
+
+
+def async_download(
+    email_address: str,
+    file_format: str,
+    funds: list[str] | None = None,
+    organisations: list[str] | None = None,
+    regions: list[str] | None = None,
+    rp_start: str | None = None,
+    rp_end: str | None = None,
+    outcome_categories: list[str] | None = None,
+):
+    """Download data, store file in S3 and send an email to the user with the download link."""
+
+    response = download(
+        file_format=file_format,
+        funds=funds,
+        organisations=organisations,
+        regions=regions,
+        rp_start=rp_start,
+        rp_end=rp_end,
+        outcome_categories=outcome_categories,
+    )
+
+    file_format_str: str = ""
+    if file_format == "xlsx":
+        file_format_str = "Microsoft Excel spreadsheet"
+    elif file_format == "json":
+        file_format_str = "JSON file"
+
+    file_size_kb = round(len(response.data) / 1024, 1)
+    if file_size_kb < 1024:
+        file_size_str: str = f"{round(file_size_kb, 1)} KB"
+    else:
+        file_size_str = f"{round(file_size_kb / 1024, 1)} MB"
+
+    # Upload the file to S3 and get presigned URL
+    file_obj = io.BytesIO(response.data)
+    bucket = Config.AWS_S3_BUCKET_FIND_DATA_FILES
+    find_service_base_url = Config.FIND_SERVICE_BASE_URL
+
+    if not bucket:
+        current_app.logger.error("AWS_S3_BUCKET_FIND_DATA_FILES is not set.")
+    elif not find_service_base_url:
+        current_app.logger.error("FIND_SERVICE_BASE_URL is not set.")
+    else:
+        file_name = f"{str(uuid.uuid4())}.{file_format}"
```

Comment:
Can we give this a nicer filename so that when the user downloads it, it's more meaningful?

Maybe just something like `fund-monitoring-data-YYYY-MM-DD-HH:MM:SS.xlsx`? Not the nicest, but better than a random selection of letters and numbers...

Reply:
File name changed to:
current_datetime = datetime.now().strftime("%Y-%m-%d-%H:%M:%S")
file_name = f"fund-monitoring-data-{current_datetime}.{file_format}"